### PR TITLE
Add telemetry for uploading failed runs

### PR DESCRIPTION
**Pull request by henrymercer**
#1393 started uploading failed runs to the code scanning API.  In preparation for rolling this out, it would be useful to gather some telemetry, particularly since things are set up to be unlikely to hard-fail, so we probably won't get any customer feedback.

- [Example run with a skipped reason](https://github.com/github/codeql-action/actions/runs/3651268514/jobs/6168362274)
- [Example run with an error](https://github.com/github/codeql-action/actions/runs/3651420972/jobs/6168671829)

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: src/init-action-post-helper.ts [hunk 1]
```diff
@@ -69,31 +100,32 @@ export async function uploadFailedSarif(
...
-      await uploadFailedSarif(config, repositoryNwo, featureEnablement, logger);
+      return await uploadFailedSarif(
+        config,
+        repositoryNwo,
+        featureEnablement,
+        logger
+      );
     } catch (e) {
       if (expectFailedSarifUpload) {
         throw new Error(
```
Comment (aeisenberg) [inline src/init-action-post-helper.ts:131]: This only happens during testing, right? Also, below.

File: src/init-action-post-helper.ts [hunk 2]
```diff
@@ -69,31 +100,32 @@ export async function uploadFailedSarif(
     logger,
     { isUnsuccessfulExecution: true }
   );
+  return {
+    upload_failed_run_raw_upload_size_bytes:
+      uploadResult?.statusReport?.raw_upload_size_bytes,
+    upload_failed_run_zipped_upload_size_bytes:
+      uploadResult?.statusReport?.zipped_upload_size_bytes,
+  };
```
Comment (aeisenberg) [inline src/init-action-post-helper.ts:108]: So, with this change, there will be two status reports in this case? One for the upload and one for logging the upload failure, I think. If so, that might lead to some confusion.

File: src/init-action-post-helper.ts (hunk 1 above)
Comment (henrymercer) [reply src/init-action-post-helper.ts:131]: Correct — `expectFailedSarifUpload` is just for testing.  There's a comment about that in the code, but you have to expand the diff to see it.

File: src/init-action-post-helper.ts (hunk 2 above)
Comment (henrymercer) [reply src/init-action-post-helper.ts:108]: There's a bit of name overloading going on here — the upload process doesn't submit a status report to the API.  `uploadResult?.statusReport` is rather a list of fields that:

- Are included in the `finish` status report sent at the end of `analyze`
- Are included in the status report sent at the end of `upload-sarif`
- (As of this PR) are included in the status report sent at the end of the post `init` Action.

So we only ever upload one status report about the failed SARIF file.

Note that if `analyze` produces a SARIF file, tries to upload that SARIF file, but uploading that SARIF fails catastrophically, we will get a `finish` status report with info about that SARIF upload attempt, and also a `init-post` status report with info about the failed SARIF file upload.

Comment (henrymercer) [reply src/init-action-post-helper.ts:108]: There's a bit of name overloading going on here — the upload process doesn't submit a status report to the API.  `uploadResult?.statusReport` is rather a list of fields that:

- Are included in the `finish` status report sent at the end of `analyze`
- Are included in the status report sent at the end of `upload-sarif`
- (As of this PR) are included in the status report sent at the end of the post `init` Action.

So we only ever upload one status report about the failed SARIF file.

Note that if `analyze` produces a SARIF file, tries to upload that SARIF file, but uploading that SARIF fails catastrophically, we will get a `finish` status report with info about that SARIF upload attempt, and also a `init-post` status report with info about the failed SARIF file upload.

Comment (aeisenberg) [reply src/init-action-post-helper.ts:108]: Ah...ok. That makes sense, but this leads to another question: why change the name from `raw_upload_size_bytes` to `upload_failed_run_raw_upload_size_bytes` and the other field too? When we upload the failed status report, we already know that the report is for a failure. Is there any need to make these fields differently named?

Comment (henrymercer) [reply src/init-action-post-helper.ts:108]: Introducing a new field is a more conservative option that reduces the risk of breaking internal queries that look at `raw_upload_size_bytes` without filtering to `action_name in { "finish", "upload-sarif" }`.  I have looked at some common places and can't find any queries that don't look at all the fields, but I don't believe there's a way to check all uses of telemetry.

I'm ambivalent about which way we go here — let me know how convinced you are here.

Comment (angelapwen) [reply src/init-action-post-helper.ts:108]: Just as another data point, I also found the rename a bit confusing (but understand the conservative approach)

File: src/init-action-post-helper.ts [hunk 3]
```diff
@@ -104,12 +136,41 @@ export async function run(
...
+  featureEnablement: FeatureEnablement,
+  logger: Logger
+) {
+  const config = await getConfig(actionsUtil.getTemporaryDirectory(), logger);
+  if (config === undefined) {
+    logger.warning(
+      "Debugging artifacts are unavailable since the 'init' Action failed before it could produce any."
+    );
+    return;
+  }
```
Comment (angelapwen) [inline src/init-action-post-helper.ts:166]: 👍 thanks!

File: src/init-action-post.ts [hunk 4]
```diff
@@ -27,22 +42,45 @@ async function runWrapper() {
...
+    "init-post",
+    "success",
+    startedAt
+  );
+  const statusReport: InitPostStatusReport = {
+    ...statusReportBase,
+    ...uploadFailedSarifResult,
+  };
+  console.log("status report");
+  console.log(statusReport);
```
Comment (angelapwen) [inline src/init-action-post.ts:82]: Are these just for debug purposes? I don't think we're logging the full status report in other instances.

File: src/init-action-post-helper.ts (hunk 2 above)
Comment (henrymercer) [reply src/init-action-post-helper.ts]: Thanks, that's helpful.  I've pushed a commit to re-use the existing fields.

File: src/init-action-post.ts (hunk 4 above)
Comment (henrymercer) [reply src/init-action-post.ts:82]: Yeah, just for debugging.  I think I removed this in a subsequent commit.

Comment (henrymercer) [reply src/init-action-post.ts:82]: Oops, these should go, thanks!

File: src/init-action-post-helper.ts [hunk 5]
```diff
@@ -16,17 +16,37 @@ import {
   getWorkflow,
 } from "./workflow";
 
+export interface UploadFailedSarifResult extends uploadLib.UploadStatusReport {
```
Comment (angelapwen) [inline src/init-action-post-helper.ts:19]: Non-blocking: just wondering why we're now extending this `UploadStatusReport`? Is it just best practice to do so when we're submitting results related to the status of an upload?

**Review by angelapwen — APPROVED** [review]
Thanks!!

File: src/init-action-post-helper.ts (hunk 5 above)
Comment (henrymercer) [reply src/init-action-post-helper.ts:19]: This is just to indicate that this result may include the `raw_upload_size_bytes` and `zipped_upload_size_bytes` fields from the `UploadStatusReport`

Comment (angelapwen) [reply src/init-action-post-helper.ts:19]: A ha, I see 👍

**Pull request merged**